# 23SP COMSC-140 Midterm

Note: there are 7 steps to this midterm (see the Midterm Requirements below). In each step, copy and paste your code from the previous step into the cell for the current step, and modify that code. Each cell will be a complete, standalone Python program.

# Math Tutor Program

Write a math tutoring program. The computer will present the user with random math questions. The user enters their guess, and the computer tells them if they're right or not. The computer also tracks the number of correct answers.

Here's an example of the program's output:

    MATH TUTOR! You'll get 4 random problems
    using four random operators.  Press ENTER to start!

    For easy level enter 1
        medium           2
        hard             3
        extreme          4
    Choice? --> 1

    Asking 4 questions:

    Question 1: 5 + 7 = 12
        Result: Correct!
    Question 2: 1 * 1 = 1
        Result: Correct!
    Question 3: 1 * 6 = 7
        Result: Incorrect; the answer was 6
    Question 4: 10 * 10 = 100
        Result: Correct!

    Tutoring complete! You got 3 of 4 correct (75.00%).
    Another set of problems? Y/N: n


# Midterm Requirements

### To compete for a max grade of 90:

Complete steps 1-5.

### To compete for a max grade of 100:

Complete steps 1-6.

### To compete for a max grade of 105:

Complete steps 1-7.



## Step 1: Write the Program Engine

We will start writing this tutoring program by creating the program engine, which is the outer loop that repeats the entire tutoring program until the user elects to quit. Create this while loop. Just have one statement in the loop for now:

    print('Example problems')

Then get the user's choice to repeat (make sure to validate this choice).

Also, set a constant QUESTIONS to 4; this will be the four questions per round that we will start with.

After this step, the user should be able to run the program as many times as they want, until they elect to quit by entering 'N' or 'n'. 

### Sample Output So Far in Step 1

Note the input validation in the Y/N entry.

    Example problems
    
    Another set of problems? Y/N: u
        ERROR: Y/N only: i
        ERROR: Y/N only: y
    Example problems
    
    Another set of problems? Y/N: n

In [1]:
QUESTIONS = 4

def ask():
    valid_reponses = ['y', 'n']
    response = 'y'
    while response != 'n':
        print('Example problems\n')
        msg = 'Another set of problems? [y/n]:'
        input_val = input(msg).lower()
        print(msg, input_val)
        response = input_val
        while not(response in valid_reponses):
            msg = 'ERROR: [y/n] only:'
            input_val = input(msg).lower()
            response = input_val
            print(msg, input_val)

ask()

Example problems

Another set of problems? [y/n]: x
ERROR: [y/n] only: x
ERROR: [y/n] only: a
ERROR: [y/n] only: n


## Step 2: Welcome Banner Function

Copy your code from the cell above and paste it into the cell below. We'll keep following that process where one cell builds on the previous cell. But keep each cell as a standalone program. Don't split your code over two cells.

Write a function 'banner()' that outputs this banner:

    MATH TUTOR! You'll get 4 problems to solve. Each problem will use
    random numbers as operands, and you'll have to perform a random
    math operation (add, subtract, multiply, divide).
    
    Press START when ready.

To press enter to start, just simply code 'input()'. This instruction will wait for enter to be pressed.

Call this function at the top of your main function.

### Sample Output So Far in Step 2

    MATH TUTOR! You'll get 4 problems to solve. Each problem will use
    random numbers as operands, and you'll have to perform a random
    math operation (add, subtract, multiply, divide).

    Press START when ready.

    Example problems
    
    Another set of problems? Y/N: n

In [6]:
QUESTIONS = 4

def ask():
    '''Asks the problems'''
    response = 'y'
    while response != 'n':
        print('Example problems\n')
        response = confirm_repeat()

def confirm_repeat() -> str:
    '''Confirms if user wants to get new set of problems'''
    valid_reponses = ['y', 'n']
    msg = 'Another set of problems? [y/n]:'
    input_val = input(msg).lower()
    print(msg, input_val)
    response = input_val
    while not(response in valid_reponses):
        msg = 'ERROR: [y/n] only:'
        input_val = input(msg).lower()
        response = input_val
        print(msg, input_val)
    return response

def banner():
    '''Prints the welcome banner'''
    print("MATH TUTOR! You'll get 4 problems to solve. Each problem will use")
    print("random numbers as operands, and you'll have to perform a random")
    print("math operation (add, subtract, multiply, divide).")
    print("")
    print("Press START when ready.")
    input()

def main():
    '''Entry point of the program'''
    banner()
    ask()

main()

MATH TUTOR! You'll get 4 problems to solve. Each problem will use
random numbers as operands, and you'll have to perform a random
math operation (add, subtract, multiply, divide).

Press START when ready.
Example problems

Another set of problems? [y/n]: y
Example problems

Another set of problems? [y/n]: n


## Step 3: Write the Question Engine

Code two more constants, MIN_RAND and MAX_RAND, and set these to 1 & 10 respectively.

Import random.

Inside your program engine, we'll now create the Question Engine, which is a for loop that runs QUESTIONS times (remember QUESTIONS is a constant we set to 4).

Replace your print('Example problems') statement with a for loop running four times. This is your question engine.

Inside this question engine, generate two random operands in the range MIN_RAND and MAX_RAND.

For now, just print out the two operands that were generated. You can use:

    print(f'{op1}, {op2}')

### Sample Output So Far in Step 3

    MATH TUTOR! You'll get 4 random problems
    using four random operators.  Press ENTER to start!

    3, 9
    4, 9
    3, 2
    1, 7
    
    Another set of problems? Y/N: n

In [8]:
import random

MIN_RAND = 1
MAX_RAND = 10
QUESTIONS = 4

def ask(repeat = True):
    '''Asks the problems'''
    if not repeat:
        return
    divider()
    for i in range(QUESTIONS):
        numbers = generate_numbers()
        print(f'{numbers[0]}, {numbers[1]}')
    print('') # spacing
    ask(confirm_repeat())

def generate_numbers() -> list[int]:
    '''Generates 2 random numbers'''
    numbers = []
    for i in range(2):
        numbers.append(random.randint(MIN_RAND, MAX_RAND))
    return numbers

def confirm_repeat() -> bool:
    '''Confirms if user wants to get new set of problems.'''
    valid_reponses = ['y', 'n']
    msg = 'Another set of problems? [y/n]:'
    input_val = input(msg).lower()
    print(msg, input_val)
    response = input_val
    while not(response in valid_reponses):
        msg = 'ERROR: [y/n] only:'
        input_val = input(msg).lower()
        response = input_val
        print(msg, input_val)
    return (response == 'y')

def banner():
    '''Prints the welcome banner'''
    print("MATH TUTOR! You'll get 4 problems to solve. Each problem will use")
    print("random numbers as operands, and you'll have to perform a random")
    print("math operation (add, subtract, multiply, divide).")
    print("")
    print("Press START when ready.")
    input()

def divider():
    print("------------------------------------------------------------------")

def main():
    '''Entry point of the program'''
    banner()
    ask()

main()

MATH TUTOR! You'll get 4 problems to solve. Each problem will use
random numbers as operands, and you'll have to perform a random
math operation (add, subtract, multiply, divide).

Press START when ready.
------------------------------------------------------------------
10, 6
4, 9
6, 9
10, 1
Another set of problems? [y/n]: y
------------------------------------------------------------------
1, 5
1, 3
4, 7
7, 2
Another set of problems? [y/n]: x
ERROR: [y/n] only: x
ERROR: [y/n] only: y
------------------------------------------------------------------
1, 4
9, 4
10, 5
3, 2
Another set of problems? [y/n]: n


## Step 4: Refine the Question Engine

"Question Engine" means the for loop running QUESTIONS times.

Now instead of outputting the operands inside the game engine, we will randomly select a math operation.

Inside your question engine, create a variable 'operation' and assign it a random number 1-3. 

    1 represents addition
    2 represents subtraction
    3 represents multiplication
    --> Note that we are not doing division yet - see Step 7 <--

Create an if/elif/else structure now. If operation is 1, calculate the result and (importantly) store this in a variable 'result' (we'll need it later). Output the math problem like this:

    3 + 9 = 12

If operation is a 2, it'd be a subtraction problem, and if operation is a 3, it'd be a multiplication problem.

### Sample Output So Far in Step 4

    MATH TUTOR! You'll get 4 random problems
    using four random operators.  Press ENTER to start!

    1 * 8 = 8
    1 * 3 = 3
    6 - 6 = 0
    4 * 2 = 8
    
    Another set of problems? Y/N: n

In [9]:
import random

MIN_RAND = 1
MAX_RAND = 10
QUESTIONS = 4

def ask(repeat = True):
    '''Asks the problems'''
    if not repeat:
        return
    divider()

    for i in range(QUESTIONS):
        numbers = get_numbers()
        operation = get_operation()
        index = operation - 1
        symbols = ['+', '-', '*']

        if operation == 1: # addition
            result = numbers[0] + numbers[1]
        elif operation == 2: # subtraction
            result = numbers[0] - numbers[1]
        else: # the only one left is multiplication
            result = numbers[0] * numbers[1]
        
        print(f'{numbers[0]} {symbols[index]} {numbers[1]} = {result}')
        
    print('') # spacing
    ask(confirm_repeat())

def get_numbers() -> list[int]:
    '''Generates 2 random numbers'''
    numbers = []
    for i in range(2):
        numbers.append(random.randint(MIN_RAND, MAX_RAND))
    return numbers

def get_operation() -> int:
    '''Generates an operation. 1 = +, 2 = -, 3 = *'''
    return random.randint(1, 3)

def confirm_repeat() -> bool:
    '''Confirms if user wants to get new set of problems.'''
    valid_reponses = ['y', 'n']
    msg = 'Another set of problems? [y/n]:'
    input_val = input(msg).lower()
    print(msg, input_val)
    response = input_val
    while not(response in valid_reponses):
        msg = 'ERROR: [y/n] only:'
        input_val = input(msg).lower()
        response = input_val
        print(msg, input_val)
    return (response == 'y')

def banner():
    '''Prints the welcome banner'''
    print("MATH TUTOR! You'll get 4 problems to solve. Each problem will use")
    print("random numbers as operands, and you'll have to perform a random")
    print("math operation (add, subtract, multiply, divide).")
    print("")
    print("Press START when ready.")
    input()

def divider():
    print("------------------------------------------------------------------")

def main():
    '''Entry point of the program'''
    banner()
    ask()

main()

MATH TUTOR! You'll get 4 problems to solve. Each problem will use
random numbers as operands, and you'll have to perform a random
math operation (add, subtract, multiply, divide).

Press START when ready.
------------------------------------------------------------------
5 - 2 = 3
2 + 6 = 8
4 - 8 = -4
4 - 2 = 2

Another set of problems? [y/n]: y
------------------------------------------------------------------
5 * 10 = 50
9 + 10 = 19
9 * 7 = 63
10 * 7 = 70

Another set of problems? [y/n]: n


## Step 5: Code User Interactivity

Paste all your code so far into the cell below.

In your program engine, add a counter variable to track the number of correct answers. Initialize this to 0.

Add user interactivity now. This coding will be within the question engine.

Update your print statements to remove the result, so the question is presented to the user like this:

    Question 1: 9 + 4 = 

Use the tag end='' i this print statement to suppress the newline character. This is so the user can enter their input on the same line.

Now write the code that allows the user to enter their answer. If the answer is correct, state so and increment the correct counter. If the answer is incorrect, state so and display the correct answer.

After the question engine but still within the program engine, display the correct question counter and percentage.

### Consider writing a process() function

Hint: in my solution, I create a function called process() that obtains the user's answer, determines if that answer is the same as the true result, and returns a 1 if correct, 0 if incorrect. I then add that return value to the counter of correct answers with: 

    count += process(result).

### Sample Output So Far in Step 5:

    MATH TUTOR! You'll get 4 random problems
    using four random operators.  Press ENTER to start!

    Asking 4 questions:

    Question 1: 8 * 6 = 48
        Result: Correct!
    Question 2: 1 - 3 = -2
        Result: Correct!
    Question 3: 7 * 3 = 22
        Result: Incorrect; the answer was 21
    Question 4: 8 + 2 = 10
        Result: Correct!

    Tutoring complete! You got 3 of 4 correct (75.00%).
    
    Another set of problems? Y/N: n


In [11]:
import random

MIN_RAND = 1
MAX_RAND = 10
QUESTIONS = 4

def ask(repeat = True):
    '''Asks the problems'''
    if not repeat:
        return
    divider()

    correct = 0
    for i in range(QUESTIONS):
        numbers = get_numbers()
        operation = get_operation() - 1 # to match array base
        symbols = ['+', '-', '*']
        operator = symbols[operation]
        correct += process(numbers[0], numbers[1], operator)
        
    correct_pct = round(100 * correct / QUESTIONS, 2)
    print(f'\nTutoring complete! You got {correct} out of {QUESTIONS} correct ({correct_pct}%)\n')
    ask(confirm_repeat())

def get_numbers() -> list[int]:
    '''Generates 2 random numbers'''
    numbers = []
    for i in range(2):
        numbers.append(random.randint(MIN_RAND, MAX_RAND))
    return numbers

def get_operation() -> int:
    '''Generates an operation. 1 = +, 2 = -, 3 = *'''
    return random.randint(1, 3)

def process(num1: int, num2: int, operator: str) -> int:
    '''Asks user to answer the question. Will return 1 if correct, 0 if incorrect'''

    result = {
        '+': num1 + num2,
        '-': num1 - num2,
        '*': num1 * num2
    }

    question = f'{num1} {operator} {num2} ='
    answer = int(input(question))
    print(question, answer)
    correct_answer = result[operator]

    if answer == correct_answer:
        print('\tResult: Correct!')
        return 1
    else:
        print(f'\tResult: Incorrect, the correct answer is {correct_answer}')
        return 0

def confirm_repeat() -> bool:
    '''Confirms if user wants to get new set of problems.'''
    valid_reponses = ['y', 'n']
    msg = 'Another set of problems? [y/n]:'
    input_val = input(msg).lower()
    print(msg, input_val)
    response = input_val
    while not(response in valid_reponses):
        msg = 'ERROR: [y/n] only:'
        input_val = input(msg).lower()
        response = input_val
        print(msg, input_val)
    return (response == 'y')

def banner():
    '''Prints the welcome banner'''
    print("MATH TUTOR! You'll get 4 problems to solve. Each problem will use")
    print("random numbers as operands, and you'll have to perform a random")
    print("math operation (add, subtract, multiply).")
    print("")
    print("Press ENTER to start.")
    input()

def divider():
    print("------------------------------------------------------------------")

def main():
    '''Entry point of the program'''
    banner()
    ask()

main()

MATH TUTOR! You'll get 4 problems to solve. Each problem will use
random numbers as operands, and you'll have to perform a random
math operation (add, subtract, multiply, divide).

Press START when ready.
------------------------------------------------------------------
2 - 8 = -6
	Result: Correct!
9 - 7 = -2
	Result: Incorrect, the correct answer is 2
5 + 9 = 14
	Result: Correct!
2 * 9 = 18
	Result: Correct!

Tutoring complete! You got 3 out of 4 correct (75.0%)

Another set of problems? [y/n]: n


## Step 6: Add Difficulty Level Functionality

At the top of your program engine, code a difficulty level:

    For easy level enter 1
        medium           2
        hard             3
        extreme          4

First, delete your current constant for MAX_RAND. We can't use this constant any more since the maximum random numbers will be changing in this step.

Set a variable 'max_rand' (this is not a constant, but rather a variable) using these values:

    If the difficulty level is Easy: max_rand is 10
    Medium: 100
    Hard: 1000
    Extreme: 10000

Then use this variable in your random number generating statements.

### Sample Output So Far in Step 6

    MATH TUTOR! You'll get 4 random problems
    using four random operators.  Press ENTER to start!

    For easy level enter 1
        medium           2
        hard             3
        extreme          4
    Choice? --> 2
    
    Asking 4 questions:

    Question 1: 78 - 82 = -4
        Result: Correct!
    Question 2: 73 * 29 = 5263
        Result: Incorrect; the answer was 2117
    Question 3: 64 + 45 = 109
        Result: Correct!
    Question 4: 9 * 38 = 342
        Result: Correct!

    Tutoring complete! You got 3 of 4 correct (75.00%).
    Another set of problems? Y/N: n

In [13]:
import random

MIN_RAND = 1
QUESTIONS = 4

def get_difficulty() -> int:
    '''Gets the difficulty level, the higher the more difficult'''
    menu = '''For easy level enter 1
    medium           2
    hard             3
    extreme          4
'''
    prompt = 'Choice:'
    print(menu)
    response = 0
    while not(response >= 1 and response <= 4):
        response = int(input(prompt))
    print(prompt, response)
    return response

def ask(level: int, repeat = True):
    '''Asks the problems'''
    if not repeat:
        return
    divider()

    correct = 0
    for i in range(QUESTIONS):
        numbers = get_numbers(level)
        operation = get_operation() - 1 # to match array base
        symbols = ['+', '-', '*']
        operator = symbols[operation]
        correct += process(numbers[0], numbers[1], operator)
        
    correct_pct = round(100 * correct / QUESTIONS, 2)
    print(f'\nTutoring complete! You got {correct} out of {QUESTIONS} correct ({correct_pct}%)\n')
    ask(level, confirm_repeat())

def get_numbers(level: int) -> list[int]:
    '''
    Generates 2 random numbers depending on difficulty level.
    1 difficulty level represents 1 power of 10.
    '''
    numbers = []
    for i in range(2):
        numbers.append(random.randint(MIN_RAND, 10 ** level))
    return numbers

def get_operation() -> int:
    '''Generates an operation. 1 = +, 2 = -, 3 = *'''
    return random.randint(1, 3)

def process(num1: int, num2: int, operator: str) -> int:
    '''Asks user to answer the question. Will return 1 if correct, 0 if incorrect'''

    result = {
        '+': num1 + num2,
        '-': num1 - num2,
        '*': num1 * num2
    }

    question = f'{num1} {operator} {num2} ='
    answer = int(input(question))
    print(question, answer)
    correct_answer = result[operator]

    if answer == correct_answer:
        print('\tResult: Correct!')
        return 1
    else:
        print(f'\tResult: Incorrect, the correct answer is {correct_answer}')
        return 0

def confirm_repeat() -> bool:
    '''Confirms if user wants to get new set of problems.'''
    valid_reponses = ['y', 'n']
    msg = 'Another set of problems? [y/n]:'
    input_val = input(msg).lower()
    print(msg, input_val)
    response = input_val
    while not(response in valid_reponses):
        msg = 'ERROR: [y/n] only:'
        input_val = input(msg).lower()
        response = input_val
        print(msg, input_val)
    return (response == 'y')

def banner():
    '''Prints the welcome banner'''
    print("MATH TUTOR! You'll get 4 problems to solve. Each problem will use")
    print("random numbers as operands, and you'll have to perform a random")
    print("math operation (add, subtract, multiply).")
    print("")
    print("Press ENTER to start.")
    input()

def divider():
    print("------------------------------------------------------------------")

def main():
    '''Entry point of the program'''
    banner()
    divider()
    ask(get_difficulty())

main()

MATH TUTOR! You'll get 4 problems to solve. Each problem will use
random numbers as operands, and you'll have to perform a random
math operation (add, subtract, multiply).

Press ENTER to start.
------------------------------------------------------------------
For easy level enter 1
    medium           2
    hard             3
    extreme          4

Choice: 2
------------------------------------------------------------------
5 + 87 = 92
	Result: Correct!
6 - 68 = -62
	Result: Correct!
6 * 100 = 600
	Result: Correct!
39 * 18 = 1
	Result: Incorrect, the correct answer is 702

Tutoring complete! You got 3 out of 4 correct (75.0%)

Another set of problems? [y/n]: n


## Step 7: Add Division Functionality

Add a fourth possibility for a random math operation: division.

The user will have to enter their answer in whole number and fractional format, as show in this correct example:

    42 / 17 = 2 8/17

You'll need to figure out how to create the string "2 8/17" based on the two operands 42 and 17. You'll want to use string concatenation to build this string result. Remember that you can concatenate various elements together in a string like this example below. This example concatenates one value with a period and the other value, such as '10.42' or '162.42'. Don't forget to convert your numbers to strings first with str(), because you can only concatenate strings and string variables:

    result = str(val1) + '.' + str(val2)
    
How can you get the figures 2, 8, and 17 in the math example above?

    num1 = 42
    num2 = 17
    
    You can derive the "2" in "2 8/17" by "num1 // num2" (integer division).

    "17" is num2.

    Derive the "8" with num1 % num2 (modulus operator)

As a reminder, the integer division operator // will return just the integer portion of a division operation. These are true statements:

    42 // 17 = 2
    50 // 11 = 4

The modulus operator % will return just the remainder portion of a division operation. These are true statements:

    14 % 5 = 4  (5 goes evenly into 14 2 times with a remainder of 4)
    16 % 5 = 1  (5 goes evenly into 16 3 times with a remainder of 1)
    22 % 2 = 0  (2 goes evenly into 22 11 times with a remainder of 0)

The user should enter the string "2 8/17" including the space. Your answer string that you build should match that format.

### Sample Output So Far in Step 7

    MATH TUTOR! You'll get 4 random problems
    using four random operators.  Press ENTER to start!

    For easy level enter 1
        medium           2
        hard             3
        extreme          4
    Choice? --> 2
    
    Asking 4 questions:

    Question 1: 30 - 7 = 1
        Result: Incorrect; the answer was 23
    Question 2: 35 - 13 = 1
        Result: Incorrect; the answer was 22
    Reminder: answer this question in fraction format.
    Question 3: 75 / 51 = 1 24/51
        Result: Correct!
    Question 4: 22 * 25 = 3
        Result: Incorrect; the answer was 550

    Tutoring complete! You got 1 of 4 correct (25.00%).
    Another set of problems? Y/N: n